filters/builtin/compress.go

```diff
@@ -0,0 +1,323 @@
+package builtin
+
+import (
+	"compress/flate"
+	"compress/gzip"
+	"errors"
+	"io"
+	"net/http"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const bufferSize = 8192
+
+type encoding struct {
+	name string
+	q    float32
+}
+
+type encodings []*encoding
+
+type compress struct {
+	mime []string
+}
+
+type encoder interface {
+	io.WriteCloser
+	Reset(io.Writer)
+}
+
+var (
+	supportedEncodings  = []string{"gzip", "deflate"}
+	unsupportedEncoding = errors.New("unsupported encoding")
+)
+
+var defaultCompressMIME = []string{
+	"text/plain",
+	"text/html",
+	"application/json",
+	"application/javascript",
+	"application/x-javascript",
+	"text/javascript",
+	"text/css",
+	"image/svg+xml",
+	"application/octet-stream",
+}
+
+var (
+	gzipPool    = &sync.Pool{}
+	deflatePool = &sync.Pool{}
+)
+
+func init() {
+	for i := 0; i < runtime.NumCPU()*4; i++ {
```

Comment:
tested with different values by binary steps, this turned out to be the best. Not sure what are the actual chances for cache misses.